mempool/src/shared_mempool/priority.rs

```diff
@@ -0,0 +1,728 @@
+        }
+
+        // Otherwise, compare by peer ping latency (the lower the better)
+        let latency_ordering = compare_ping_latency(monitoring_metadata_a, monitoring_metadata_b);
+        if !latency_ordering.is_eq() {
+            return latency_ordering; // Only return if it's not equal
+        }
+
+        // Otherwise, simply hash the peer ID and compare the hashes.
+        // In practice, this should be relatively rare.
+        let hash_a = self.hash_peer_id(network_id_a);
+        let hash_b = self.hash_peer_id(network_id_b);
+        hash_a.cmp(&hash_b)
+    }
+
+    /// Stable within a mempool instance but random between instances
+    fn hash_peer_id(&self, peer_network_id: &PeerNetworkId) -> u64 {
+        let mut hasher = self.random_state.build_hasher();
+        hasher.write(peer_network_id.peer_id().as_ref());
+        hasher.finish()
+    }
+}
+
+/// A simple struct to hold state for peer prioritization
+#[derive(Clone, Debug)]
+pub struct PrioritizedPeersState {
+    // The current mempool configuration
+    mempool_config: MempoolConfig,
+
+    // The current list of prioritized peers
+    prioritized_peers: Arc<RwLock<Vec<PeerNetworkId>>>,
+
+    // The comparator used to prioritize peers
+    peer_comparator: PrioritizedPeersComparator,
+
+    // Whether ping latencies were observed for all peers
+    observed_all_ping_latencies: bool,
+
+    // The last time peer priorities were updated
+    last_peer_priority_update: Option<Instant>,
+
+    // The time service used to fetch timestamps
+    time_service: TimeService,
+}
+
+impl PrioritizedPeersState {
+    pub fn new(mempool_config: MempoolConfig, time_service: TimeService) -> Self {
+        Self {
+            mempool_config,
+            prioritized_peers: Arc::new(RwLock::new(Vec::new())),
+            peer_comparator: PrioritizedPeersComparator::new(),
+            observed_all_ping_latencies: false,
+            last_peer_priority_update: None,
+            time_service,
+        }
+    }
+
+    /// Returns the priority of the given peer. The lower the
+    /// value, the higher the priority.
+    pub fn get_peer_priority(&self, peer_network_id: &PeerNetworkId) -> usize {
+        let prioritized_peers = self.prioritized_peers.read();
+        prioritized_peers
+            .iter()
+            .find_position(|peer| *peer == peer_network_id)
+            .map_or(usize::MAX, |(position, _)| position)
+    }
+
+    /// Returns true iff the prioritized peers list is ready for another update.
+    /// This is based on the last time the prioritized peers were updated, and if
+    /// ping latencies were observed for all peers in the last update.
+    pub fn ready_for_update(&self, peers_changed: bool) -> bool {
+        // If our peers have changed, or we haven't observed ping latencies
+        // for all peers yet, we should update the prioritized peers again.
+        if peers_changed || !self.observed_all_ping_latencies {
```

Comment:
Can you explain the reasoning behind updating this if we have not observed ping latency for all peers?

Reply:
I see, makes sense. It might be better to provide this context with comment in the code.